### PR TITLE
Addition of Rank-Stabilized LoRA parameter

### PR DESCRIPTION
**Pull request by CoffeeVampir3**
Concept from: https://huggingface.co/blog/damjan-k/rslora

Adds rs-lora parameter to create_network.

Adds rs-lora scaling to the following networks:
glokr
glora
lilora
locon
loha
lokr

For all networks, the change is the same:
```python
self.scale = alpha / lora_dim
```
Becomes
```python
r_factor = lora_dim
if self.rs_lora:
    r_factor = math.sqrt(r_factor)

self.scale = alpha / r_factor
```

Training via dora showed that their were still potentially issues with this scaling factor, the addition of the sqrt term does experimentally seem to allow larger rank networks. The implementation is currently a bit strange, I wanted to do things in the least obtrusive way possible for clarity. Let me know if you'd like the code to be reorganized or refactored to decouple the lora_dim and r_factors.

Testing was carried out with locon and loha, I am currently performing ablations with the other flavors.

### CODE REVIEW

**Comment by KohakuBlueleaf** [comment]
Plz put import of built-in modules (math in this PR) before all the third party library

```
import math

import torch
....
```

**Comment by KohakuBlueleaf** [comment]
I change the PR base to dev branch
plz resolve the conflicts thx

**Comment by CoffeeVampir3** [comment]
Merged everything, looks good. Syncing my fork and going to run some tests with the imbalanced lora as well :+1:

**Comment by rockerBOO** [comment]
Additional context to not need to look in the code `rs_lora=True` is the argument. If using Kohya set the `--network_args "rs_lora=True"`

**Comment by elias-gaeros** [comment]
This seems to introduce a train-test gap: most implementations won't read the `rs_lora` flag in the metadata and will keep scaling the weights by `alpha / dim` resulting in a scale `sqrt(dim)` times lower than the one used during training.
I think it would be  better to export the lora with `exported_alpha = alpha * sqrt(dim)` so that loaders use a scale of `exported_alpha / dim = alpha / sqrt(dim)`, reproducing the scaling at train time.

**Comment by KohakuBlueleaf** [comment]
> @CoffeeVampir3 This seems to introduce a train-test gap: most implementations won't read the `rs_lora` flag in the metadata and will keep scaling the weights by `alpha / dim` resulting in a scale `sqrt(dim)` times lower than the one used during training. I think it would be better to export the lora with `exported_alpha = alpha * sqrt(dim)` so that loaders use a scale of `exported_alpha / dim = alpha / sqrt(dim)`, reproducing the scaling at train time.

known issue for a long time (and I just forgot to push the fix)
